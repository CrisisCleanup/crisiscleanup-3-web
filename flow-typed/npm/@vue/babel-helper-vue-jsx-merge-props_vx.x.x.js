// flow-typed signature: d7a23c53f4091ef2bf1df8afa5eb43c4
// flow-typed version: <<STUB>>/@vue/babel-helper-vue-jsx-merge-props_v^1.0.0/flow_v0.131.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@vue/babel-helper-vue-jsx-merge-props'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@vue/babel-helper-vue-jsx-merge-props' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@vue/babel-helper-vue-jsx-merge-props/dist/helper' {
  declare module.exports: any;
}

// Filename aliases
declare module '@vue/babel-helper-vue-jsx-merge-props/dist/helper.js' {
  declare module.exports: $Exports<'@vue/babel-helper-vue-jsx-merge-props/dist/helper'>;
}
