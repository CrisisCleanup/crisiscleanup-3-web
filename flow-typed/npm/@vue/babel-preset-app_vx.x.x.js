// flow-typed signature: 6ba91a92845d9a556a5bcdcc4f033fed
// flow-typed version: <<STUB>>/@vue/babel-preset-app_v^4.3.1/flow_v0.129.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@vue/babel-preset-app'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@vue/babel-preset-app' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@vue/babel-preset-app/polyfillsPlugin' {
  declare module.exports: any;
}

// Filename aliases
declare module '@vue/babel-preset-app/index' {
  declare module.exports: $Exports<'@vue/babel-preset-app'>;
}
declare module '@vue/babel-preset-app/index.js' {
  declare module.exports: $Exports<'@vue/babel-preset-app'>;
}
declare module '@vue/babel-preset-app/polyfillsPlugin.js' {
  declare module.exports: $Exports<'@vue/babel-preset-app/polyfillsPlugin'>;
}
