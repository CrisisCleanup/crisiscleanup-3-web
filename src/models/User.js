import { AuthService } from '@/services/auth.service';
import Language from '@/models/Language';
import Role from '@/models/Role';
import moment from 'moment';
import CCUModel from '@/models/model';

export default class User extends CCUModel {
  static entity = 'users';

  static fields() {
    return {
      id: this.attr(),
      first_name: this.string(''),
      last_name: this.string(''),
      email: this.string(''),
      mobile: this.string(''),
      roles: this.attr(null),
      active_roles: this.attr(null),
      files: this.attr(null),
      organization: this.attr(null),
      states: this.attr({}),
      preferences: this.attr({}),
      permissions: this.attr({}),
      beta_features: this.attr({}),
      primary_language: this.attr(null),
      secondary_language: this.attr(null),
      accepted_terms_timestamp: this.attr(null),
      accepted_terms: this.attr(null),
      social: this.attr({}),
      referring_user: this.attr({}),
      lineage: this.attr([]),
    };
  }

  get profilePictureUrl() {
    if (this.files && this.files.length) {
      const profilePictures = this.files.filter(
        (file) => file.file_type_t === 'fileTypes.user_profile_picture',
      );
      if (profilePictures.length) {
        return profilePictures[0].url;
      }
    }
    return `https://api.adorable.io/avatars/285/ccu-user-${this.id}.png`;
  }

  get currentRole() {
    return Role.query().whereIdIn(this.active_roles).get()[0];
  }

  get referringUser() {
    return User.find(this.referring_user);
  }

  get languages() {
    const languageList = [];
    if (this.primary_language) {
      languageList.push(Language.find(this.primary_language));
    }
    if (this.secondary_language) {
      languageList.push(Language.find(this.secondary_language));
    }
    return languageList;
  }

  get notificationSettings() {
    const settings = {
      has_notifications: false,
    };
    if (this.preferences && this.preferences.notification_settings) {
      return this.preferences.notification_settings;
    }
    return settings;
  }

  get full_name() {
    return `${this.first_name} ${this.last_name}`;
  }

  get facebook() {
    return this.social && this.social.facebook;
  }

  get twitter() {
    return this.social && this.social.twitter;
  }

  get isAdmin() {
    return this.active_roles.includes(1);
  }

  getStatesForIncident(incidentId, fallback = true) {
    if (
      this.states &&
      this.states.incidents &&
      this.states.incidents[incidentId]
    ) {
      return this.states.incidents[incidentId];
    }
    if (fallback) {
      return this.states;
    }
    return null;
  }

  static apiConfig = {
    actions: {
      login(email, password) {
        return this.post(
          `/api-token-auth`,
          {
            email,
            password,
          },
          { save: false },
        );
      },
      inviteUser(email, organization = null) {
        const data = {
          invitee_email: email,
        };
        if (organization) {
          data.organization = organization;
        }
        return this.post(`/invitations`, data, { save: false });
      },
      acceptInvite({ token, first_name, last_name, password, mobile, title }) {
        return this.post(
          `/invitations/accept`,
          {
            invitation_token: token,
            first_name,
            last_name,
            password,
            mobile,
            title,
          },
          { save: false },
        );
      },
      orphan(id) {
        return this.patch(`/users/${id}/orphan`);
      },
      addFile(id, file, type) {
        return this.post(
          `/users/${id}/files`,
          {
            file,
            type_t: type,
          },
          { save: false },
        );
      },
      deleteFile(id, file) {
        return this.delete(
          `/users/${id}/files`,
          {
            data: { file },
          },
          { save: false },
        );
      },
      async updateUserState(globalStates, incidentStates, reload = false) {
        /* Update user states JSON with new states.

           To be backwards-compatible with clients without per-incident states,
           update top-level with both globalStates and incidentStates
           and then update state for current incident with incidentStates.
        */
        let currentUser = User.find(AuthService.getUser().user_claims.id);
        const states = currentUser.states || {};
        const currentIncident = states.incident;
        let updatedStates = {
          ...states,
          ...globalStates,
          ...incidentStates,
        };
        let updatedIncidentStates = states.incidents || {};
        if (incidentStates) {
          updatedIncidentStates = {
            ...updatedIncidentStates,
            [currentIncident]: incidentStates,
          };
        }
        updatedStates = {
          ...updatedStates,
          ...{ incidents: updatedIncidentStates },
        };
        await User.update({
          where: currentUser.id,
          data: {
            states: updatedStates,
          },
        });
        currentUser = User.find(AuthService.getUser().user_claims.id);
        await this.patch(
          `/users/${currentUser.id}`,
          {
            states,
          },
          { save: false },
        );
        if (reload) {
          await this.get('/users/me', {});
        }
      },
      async updateUserPreferences(preferences, reload = false) {
        const currentUser = User.find(AuthService.getUser().user_claims.id);
        const newPreferences = {
          ...currentUser.preferences,
          ...preferences,
        };
        await this.patch(
          `/users/${currentUser.id}`,
          {
            preferences: newPreferences,
          },
          { save: false },
        );
        if (reload) {
          await this.get('/users/me', {});
        }
      },
      async acceptTerms() {
        const currentUser = User.find(AuthService.getUser().user_claims.id);
        await this.patch(`/users/${currentUser.id}`, {
          accepted_terms: true,
          accepted_terms_timestamp: moment().toISOString(),
        });
      },
    },
  };
}
