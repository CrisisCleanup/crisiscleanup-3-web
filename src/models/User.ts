import moment from 'moment';
import Bowser from 'bowser';
import * as Sentry from '@sentry/browser';
import { useRouter } from 'vue-router';
import { Model } from '@vuex-orm/core';
import type { Config } from '@vuex-orm/plugin-axios';
import { AuthService } from '../services/auth.service';
import Language from './Language';
import Role from './Role';
import CCUModel from '@/models/model';

export default class User extends CCUModel<User> {
  static entity = 'users';

  id!: string;

  primary_language!: string;

  secondary_language!: string;

  organization!: any;

  active_roles!: any[];

  referring_user!: string;

  preferences!: Record<string, any>;

  first_name!: string;

  last_name!: string;

  email!: string;

  mobile!: string;

  accepted_terms_timestamp!: string;

  states!: Record<string, any>;

  social!: Record<string, any>;

  files!: any[];

  lineage!: any[];

  permissions!: Record<string, boolean>;

  beta_features!: string[];

  static fields() {
    return {
      id: this.attr(''),
      first_name: this.string(''),
      last_name: this.string(''),
      email: this.string(''),
      mobile: this.string(''),
      roles: this.attr(null),
      active_roles: this.attr(null),
      files: this.attr(null),
      organization: this.attr(null),
      states: this.attr({}),
      preferences: this.attr({}),
      permissions: this.attr({}),
      beta_features: this.attr({}),
      primary_language: this.attr(null),
      secondary_language: this.attr(null),
      accepted_terms_timestamp: this.attr(null),
      accepted_terms: this.attr(null),
      social: this.attr({}),
      referring_user: this.attr({}),
      lineage: this.attr([]),
    };
  }

  static afterUpdate(model: User) {
    if (model.id === User.store().getters['auth/userId']) {
      AuthService.updateUser(model.$toJson());
      Sentry.setUser(model.$toJson());
      Sentry.setContext('user_states', model.states);
      Sentry.setContext('user_preferences', model.preferences);
      // User.store().commit("auth/setAcl", useRouter());
    }
  }

  get hasProfilePicture() {
    if (this.files && this.files.length > 0) {
      const profilePictures = this.files.filter(
        (file) => file.file_type_t === 'fileTypes.user_profile_picture',
      );
      return profilePictures.length;
    }

    return false;
  }

  get profilePictureUrl() {
    if (this.files && this.files.length > 0) {
      const profilePictures = this.files.filter(
        (file) => file.file_type_t === 'fileTypes.user_profile_picture',
      );
      if (profilePictures.length > 0) {
        return profilePictures[0].large_thumbnail_url;
      }
    }

    return `https://avatars.dicebear.com/api/bottts/${this.full_name}.svg`;
  }

  get currentRole() {
    return Role.query().whereIdIn(this.active_roles).get()[0];
  }

  get referringUser() {
    return User.find(this.referring_user);
  }

  get languages() {
    const languageList: any[] = [];
    if (this.primary_language) {
      languageList.push(Language.find(this.primary_language));
    }

    if (this.secondary_language) {
      languageList.push(Language.find(this.secondary_language));
    }

    return languageList;
  }

  get languageIds() {
    const languageList: any[] = [];
    if (this.primary_language) {
      languageList.push(this.primary_language);
    }

    if (this.secondary_language) {
      languageList.push(this.secondary_language);
    }

    return languageList;
  }

  get notificationSettings() {
    const settings = {
      has_notifications: false,
    };
    if (this.preferences && this.preferences.notification_settings) {
      return this.preferences.notification_settings;
    }

    return settings;
  }

  get full_name() {
    return `${this.first_name} ${this.last_name}`;
  }

  get facebook() {
    return this.social && this.social.facebook;
  }

  get twitter() {
    return this.social && this.social.twitter;
  }

  get isAdmin() {
    return this.active_roles.includes(1);
  }

  get isPrimaryContact() {
    return this.active_roles.includes(3);
  }

  getStatesForIncident(incidentId: string, fallback = true) {
    if (
      this.states &&
      this.states.incidents &&
      this.states.incidents[incidentId]
    ) {
      return this.states.incidents[incidentId];
    }

    if (fallback) {
      return this.states;
    }

    return null;
  }

  static apiConfig: Config = {
    actions: {
      login(email: string, password: string) {
        return this.post(
          `/api-token-auth`,
          {
            email,
            password,
          },
          { save: false },
        );
      },
      inviteUser(email: string, organization = null) {
        const data: Record<string, any> = {
          invitee_email: email,
        };
        if (organization) {
          data.organization = organization;
        }

        return this.post(`/invitations`, data, { save: false });
      },
      acceptInvite({
        token,
        first_name,
        last_name,
        password,
        mobile,
        title,
      }: Record<string, any>) {
        return this.post(
          `/invitations/accept`,
          {
            invitation_token: token,
            first_name,
            last_name,
            password,
            mobile,
            title,
          },
          { save: false },
        );
      },
      orphan(id: string) {
        return this.patch(`/users/${id}/orphan`);
      },
      addFile(id: string, file: string, type: string) {
        return this.post(
          `/users/${id}/files`,
          {
            file,
            type_t: type,
          },
          { save: false },
        );
      },
      deleteFile(id: string, file: string) {
        return this.delete(
          `/users/${id}/files`,
          {
            data: { file },
          },
          { save: false },
        );
      },
      async clearUserStates() {
        const currentUser = User.find(User.store().getters['auth/userId']);
        await this.patch(
          `/users/${currentUser?.id}`,
          {
            states: {},
          },
          { save: false },
        );
        await this.get('/users/me');
      },
      async clearUserPreferences() {
        const currentUser = User.find(User.store().getters['auth/userId']);
        await this.patch(
          `/users/${currentUser?.id}`,
          {
            preferences: {},
          },
          { save: false },
        );
        await this.get('/users/me');
      },
      async updateUserState(
        globalStates: Record<string, any>,
        incidentStates: Record<string, any>,
        reload = false,
      ) {
        /* Update user states JSON with new states.

           To be backwards-compatible with clients without per-incident states,
           update top-level with both globalStates and incidentStates
           and then update state for current incident with incidentStates.
        */
        let currentUser = User.find(User.store().getters['auth/userId']);

        if (!currentUser) {
          return;
        }

        const states = (currentUser && currentUser.states) || {};
        const currentIncident = states.incident;
        let updatedStates = {
          ...states,
          ...globalStates,
          ...incidentStates,
        };
        let updatedIncidentStates = states.incidents || {};
        if (incidentStates) {
          updatedIncidentStates = {
            ...updatedIncidentStates,
            [currentIncident]: incidentStates,
          };
        }

        updatedStates = {
          ...updatedStates,
          incidents: updatedIncidentStates,
          userAgent: Bowser.parse(window.navigator.userAgent),
        };
        await User.update({
          where: currentUser.id,
          data: {
            states: updatedStates,
          },
        });
        currentUser = User.find(AuthService.getUser().user_claims.id);
        await this.patch(
          `/users/${currentUser?.id}`,
          {
            states: updatedStates,
          },
          { save: false },
        );
        if (reload) {
          await this.get('/users/me');
        }
      },
      async updateUserPreferences(
        preferences: Record<string, any>,
        reload = false,
      ) {
        const currentUser = User.find(AuthService.getUser().user_claims.id);
        if (!currentUser) {
          return;
        }

        const newPreferences = {
          ...currentUser.preferences,
          ...preferences,
        };
        await this.patch(
          `/users/${currentUser.id}`,
          {
            preferences: newPreferences,
          },
          { save: false },
        );
        if (reload) {
          await this.get('/users/me', {});
        }
      },
      async acceptTerms() {
        const currentUser = User.find(AuthService.getUser().user_claims.id);
        await this.patch(`/users/${currentUser?.id}`, {
          accepted_terms: true,
          accepted_terms_timestamp: moment().toISOString(),
        });
      },
    },
  };
}
