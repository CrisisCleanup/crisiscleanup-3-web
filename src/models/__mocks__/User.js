import { MockUsers, mockModel } from '@/utils/testing';

export default mockModel(MockUsers, {
  api: {
    inviteUser: jest.fn(),
  },
});
