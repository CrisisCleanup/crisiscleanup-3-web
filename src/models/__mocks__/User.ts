import { MockUsers, mockModel } from '@/utils/testing';

// @ts-ignore
export default mockModel(MockUsers, {
  api: {
    inviteUser: jest.fn(),
  },
});
