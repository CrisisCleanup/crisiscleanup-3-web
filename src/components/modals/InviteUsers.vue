<template>
  <div>
    <base-button
      :text="$t('usersVue.invite_new_user')"
      :alt="$t('usersVue.invite_new_user')"
      variant="solid"
      class="mx-1 px-3 py-1"
      :action="
        () => {
          showInviteModal = true;
        }
      "
    />
    <modal
      v-if="showInviteModal"
      modal-classes="bg-white max-w-2xl shadow"
      :title="$t('usersVue.invite_user')"
      closeable
      @close="
        () => {
          emails = [];
          showInviteModal = false;
        }
      "
    >
      <div class="text-justify flex flex-col p-3 justify-center">
        <div class="my-3">
          {{ $t('inviteTeammates.invite_teammates_instructions') }}
        </div>
        <div class="mb-4">
          <tag-input
            v-model="emails"
            v-model:tags="usersToInvite"
            :placeholder="$t('usersVue.emails')"
            :validation="validation"
            :add-on-key="[13, 32, ',']"
            :separators="[';', ',', ', ']"
            @tags-changed="(newTags) => (usersToInvite = newTags)"
          />
        </div>
        <div v-if="isAdmin || currentOrganization.affiliates.length > 1">
          <OrganizationSearchInput
            @selectedOrganization="(id) => (selectedOrganization = id)"
            class="w-108"
            :allowed-organization-ids="currentOrganization.affiliates"
            :is-admin="isAdmin"
          />
        </div>
      </div>
      <div slot="footer" class="p-3 flex justify-end">
        <base-button
          :text="$t('actions.cancel')"
          :alt="$t('actions.cancel')"
          class="ml-2 p-3 px-6 mr-1 text-xs border border-black"
          :action="
            () => {
              showInviteModal = false;
            }
          "
        />
        <base-button
          variant="solid"
          :action="() => inviteUsers()"
          :text="$t('actions.submit_invites')"
          :alt="$t('actions.submit_invites')"
          class="ml-2 p-3 px-6 text-xs"
        />
      </div>
    </modal>
  </div>
</template>
<script>
import _ from 'lodash';
import Multiselect from '@vueform/multiselect';
import { computed, ref } from 'vue';
import { useToast } from 'vue-toastification';
import { useI18n } from 'vue-i18n';
import { createTags } from '@sipec/vue3-tags-input';
import User from '../../models/User';
import Organization from '../../models/Organization';
import OrganizationSearchInput from '../OrganizationSearchInput.vue';
import { getErrorMessage } from '../../utils/errors';
import useCurrentUser from '../../hooks/useCurrentUser';

const EMAIL_REGEX = /[^@ \t\r\n]+@[^@ \t\r\n]+\.[^@ \t\r\n]+/;

export default {
  name: 'InviteUsers',
  components: { OrganizationSearchInput, Multiselect },
  props: {
    isAdmin: {
      type: Boolean,
      default: false,
    },
  },
  setup() {
    const $toasted = useToast();
    const { t } = useI18n();

    const validation = [
      {
        classes: 'email',
        rule: /[\w.!#$%&â€™*+/=?^`{|}~-]+@[a-zA-Z\d-]+(?:\.[a-zA-Z\d-]+)*/,
        disableAdd: true,
      },
    ];
    const emails = ref('');
    const usersToInvite = ref([]);
    const multiselect = ref(null);
    const showInviteModal = ref(false);
    const selectedOrganization = ref(null);
    const organizationResults = ref([]);
    const { currentUser } = useCurrentUser();
    const currentOrganization = computed(() =>
      Organization.find(currentUser?.organization?.id),
    );

    async function onOrganizationSearch(value) {
      const results = await Organization.api().get(
        `/organizations?search=${value}&limit=10&fields=id,name&is_active=true`,
        {
          dataKey: 'results',
        },
      );
      organizationResults.value = results.entities.organizations;
    }
    const inviteUsers = async () => {
      let tags = _.defaultTo([...usersToInvite.value], []);
      try {
        if (emails.value) {
          const emailList = emails.value.match(EMAIL_REGEX);
          let extTags = _.attempt(createTags, emailList);
          if (_.isError(extTags)) {
            extTags = [];
          }
          tags = _.uniqBy([...tags, ...extTags], 'text');
        }
        if (_.isEmpty(tags)) {
          await $toasted.error(t('inviteTeammates.provide_valid_email'));
          return;
        }
        const emailsGroup = tags.map((value) => value.text);
        await Promise.all(
          emailsGroup.map((email) =>
            User.api().inviteUser(email, selectedOrganization.value),
          ),
        );
        await $toasted.success(t('inviteTeammates.invites_sent_success'));
        showInviteModal.value = false;
        usersToInvite.value = [];
      } catch (error) {
        await $toasted.error(getErrorMessage(error));
      }
    }

    return {
      validation,
      emails,
      showInviteModal,
      currentUser,
      currentOrganization,
      onOrganizationSearch,
      selectedOrganization,
      organizationResults,
      inviteUsers,
      multiselect,
      usersToInvite,
    };
  },
};
</script>

<style>
.vue-tags-input {
  @apply h-auto w-full mb-2;
}
.vue-tags-input .ti-input {
  @apply h-auto;
}
</style>
