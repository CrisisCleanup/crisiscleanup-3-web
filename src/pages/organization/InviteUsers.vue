<template>
  <div>
    <base-button
      :text="$t('usersVue.invite_new_user')"
      variant="solid"
      class="px-3 py-1"
      :action="
        () => {
          showInviteModal = true;
        }
      "
    />
    <modal
      v-if="showInviteModal"
      modal-classes="bg-white max-w-2xl shadow"
      :title="$t('usersVue.invite_user')"
      closeable
      @close="
        () => {
          usersToInvite = [];
          showInviteModal = false;
        }
      "
    >
      <div class="text-justify flex flex-col p-3 justify-center">
        <div class="my-3">
          {{ $t('inviteTeammates.invite_teammates_instructions') }}
        </div>
        <div>
          <tag-input
            v-model="emails"
            :tags="usersToInvite"
            :placeholder="$t('usersVue.emails')"
            :add-on-key="[13, ',']"
            @before-adding-tag="
              (obj) => {
                let emailMatch = getEmailMatch(obj.tag.text);
                if (emailMatch) {
                  obj.tag.text = emailMatch;
                  obj.addTag();
                }
              }
            "
            :separators="[';', ',', ', ']"
            @tags-changed="(newTags) => (usersToInvite = newTags)"
          />
        </div>
        <div v-if="isAdmin">
          <OrganizationSearchInput
            @selectedOrganization="selectedOrganization = $event.id"
            class="w-108"
          />
        </div>
      </div>
      <div slot="footer" class="p-3 flex justify-end">
        <base-button
          :text="$t('actions.cancel')"
          class="ml-2 p-3 px-6 mr-1 text-xs border border-black"
          :action="
            () => {
              showInviteModal = false;
            }
          "
        />
        <base-button
          variant="solid"
          :action="inviteUsers"
          :text="$t('actions.submit_invites')"
          class="ml-2 p-3 px-6 text-xs"
        />
      </div>
    </modal>
  </div>
</template>
<script>
import User from '@/models/User';
import Organization from '@/models/Organization';
import { createTag } from '@johmun/vue-tags-input';
import OrganizationSearchInput from '@/components/OrganizationSearchInput';
import { getErrorMessage } from '../../utils/errors';

export default {
  name: 'InviteUsers',
  components: { OrganizationSearchInput },
  props: {
    isAdmin: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      emails: '',
      usersToInvite: [],
      showInviteModal: false,
      selectedOrganization: null,
      organizationResults: [],
    };
  },
  methods: {
    getEmailMatch(text) {
      const emailMatch = text
        .toLowerCase()
        .match(
          /[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*/,
        );
      if (emailMatch) {
        return emailMatch[0];
      }
      return null;
    },
    async onOrganizationSearch(value) {
      const results = await Organization.api().get(
        `/organizations?search=${value}&limit=10&fields=id,name`,
        {
          dataKey: 'results',
        },
      );
      this.organizationResults = results.entities.organizations;
    },
    async inviteUsers() {
      try {
        if (this.emails) {
          const email = createTag(this.getEmailMatch(this.emails));
          if (email) {
            this.usersToInvite.push(email);
          }
        }
        const emails = this.usersToInvite.map((value) => value.text);
        await Promise.all(
          emails.map((email) =>
            User.api().inviteUser(email, this.selectedOrganization),
          ),
        );
        await this.$toasted.success(
          this.$t('inviteTeammates.invites_sent_success'),
        );
        this.showInviteModal = false;
        this.usersToInvite = [];
      } catch (error) {
        await this.$toasted.error(getErrorMessage(error));
      }
    },
  },
};
</script>

<style>
.vue-tags-input {
  @apply h-8 w-108 mb-2;
}

.vue-tags-input .ti-input {
  @apply h-8;
}
</style>
