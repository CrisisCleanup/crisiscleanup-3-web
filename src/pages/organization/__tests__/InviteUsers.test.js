import { mount } from '@vue/test-utils';
import User from '@/models/User';
import InviteUsers from '../InviteUsers.vue';

jest.mock('@/models/User');

const mocks = {
  $t: (key) => key,
  $toasted: {
    error: jest.fn(),
  },
};

const mountWithOptions = ({ props, data, ...opts } = {}) =>
  mount(InviteUsers, {
    stubs: ['base-button', 'tag-input', 'modal'],
    propsData: {
      admin: false,
      ...props,
    },
    data() {
      return {
        ...data,
      };
    },
    mocks,
    ...opts,
  });

describe('InviteUsers', () => {
  beforeEach(() => User.api().inviteUser.mockClear());

  it('should not log any errors', () => {
    const spy = jest.spyOn(global.console, 'error');
    mountWithOptions();
    expect(spy).not.toHaveBeenCalled();
  });

  it('should convert emails to tags on submit', async () => {
    const wrapper = mountWithOptions({
      data: { showInviteModal: true, emails: 'example@example.com' },
    });
    await wrapper.vm.inviteUsers();
    expect(User.api().inviteUser.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "example@example.com",
          null,
        ],
      ]
    `);
  });

  it('should invite all tags', async () => {
    const wrapper = mountWithOptions({
      data: {
        showInviteModal: true,
        usersToInvite: [{ text: 'one@email.com' }, { text: 'two@email.com' }],
        emails: '',
      },
    });
    await wrapper.vm.inviteUsers();
    expect(User.api().inviteUser.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "one@email.com",
          null,
        ],
        Array [
          "two@email.com",
          null,
        ],
      ]
    `);
  });

  it('should combine untagged emails with tags', async () => {
    const wrapper = mountWithOptions({
      data: {
        showInviteModal: true,
        usersToInvite: [{ text: 'one@email.com' }, { text: 'two@email.com' }],
        emails: 'three@email.com',
      },
    });
    await wrapper.vm.inviteUsers();
    expect(User.api().inviteUser.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "one@email.com",
          null,
        ],
        Array [
          "two@email.com",
          null,
        ],
        Array [
          "three@email.com",
          null,
        ],
      ]
    `);
  });

  it('should not combine invalid emails with tags', async () => {
    const wrapper = mountWithOptions({
      data: {
        showInviteModal: true,
        usersToInvite: [{ text: 'one@email.com' }, { text: 'two@email.com' }],
        emails: 'three',
      },
    });
    await wrapper.vm.inviteUsers();
    expect(User.api().inviteUser.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "one@email.com",
          null,
        ],
        Array [
          "two@email.com",
          null,
        ],
      ]
    `);
  });

  it('should show error without any emails', async () => {
    const wrapper = mountWithOptions({
      data: {
        showInviteModal: true,
        usersToInvite: [],
        emails: '',
      },
    });
    await wrapper.vm.inviteUsers();
    expect(User.api().inviteUser).not.toHaveBeenCalled();
    expect(wrapper.vm.$toasted.error).toHaveBeenCalled();
  });
});
