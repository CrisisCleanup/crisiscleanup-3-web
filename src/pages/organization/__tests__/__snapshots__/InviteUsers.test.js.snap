// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InviteUsers should match snapshot 1`] = `
<div>
  <base-button-stub
    action="function () {
        /* istanbul ignore next */
        cov_2ohn4qljyu.f[7]++;
        cov_2ohn4qljyu.s[51]++;
        _vm.showInviteModal = true;
      }"
    alt="usersVue.invite_new_user"
    class="px-3 py-1"
    text="usersVue.invite_new_user"
    variant="solid"
  />
   
  <modal-stub
    closeable=""
    modal-classes="bg-white max-w-2xl shadow"
    title="usersVue.invite_user"
  >
    <div
      class="text-justify flex flex-col p-3 justify-center"
    >
      <div
        class="my-3"
      >
        
        inviteTeammates.invite_teammates_instructions
      
      </div>
       
      <div>
        <tag-input-stub
          add-on-key="13,32,,"
          placeholder="usersVue.emails"
          separators=";,,,, "
          tags=""
          validation="[object Object]"
          value=""
        />
      </div>
       
      <!---->
    </div>
     
    <div
      class="p-3 flex justify-end"
    >
      <base-button-stub
        action="function () {
        /* istanbul ignore next */
        cov_2ohn4qljyu.f[13]++;
        cov_2ohn4qljyu.s[58]++;
        _vm.showInviteModal = false;
      }"
        alt="actions.cancel"
        class="ml-2 p-3 px-6 mr-1 text-xs border border-black"
        text="actions.cancel"
      />
       
      <base-button-stub
        action="function () {
        /* istanbul ignore next */
        cov_2ohn4qljyu.f[14]++;
        cov_2ohn4qljyu.s[59]++;
        return _vm.inviteUsers();
      }"
        alt="actions.submit_invites"
        class="ml-2 p-3 px-6 text-xs"
        text="actions.submit_invites"
        variant="solid"
      />
    </div>
  </modal-stub>
</div>
`;
